src/librustc_mir/build/matches/mod.rs

```diff
@@ -443,42 +443,83 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
+    ///
+    /// - If the outcome is that `x` is true, candidates 0 and 2
+    /// - If the outcome is that `x` is false, candidates 1 and 2
     ///
-    /// - If outcome is that `x == "foo"` (one candidate, derived from `x @ "foo"`)
-    /// - If outcome is that `x != "foo"` (empty list of candidates)
-    /// - Otherwise (three candidates, `x @ "bar"`, `x @ "baz"`, `x @
-    ///   _`). Here we have the invariant that everything in the
-    ///   otherwise list is of **lower priority** than the stuff in the
-    ///   other lists.
+    /// Here, the traditional "decision tree" method would generate 2
+    /// separate code-paths for the 2 separate cases.
     ///
-    /// So we'll compile the test. For each outcome of the test, we
-    /// recursively call `match_candidates` with the corresponding set
-    /// of candidates. But note that this set is now inexhaustive: for
-    /// example, in the case where the test returns false, there are
-    /// NO candidates, even though there is stll a value to be
-    /// matched. So we'll collect the return values from
-    /// `match_candidates`, which are the blocks where control-flow
-    /// goes if none of the candidates matched. At this point, we can
-    /// continue with the "otherwise" list.
+    /// In some cases, this duplication can create an exponential amount of
+    /// code. This is most easily seen by noticing that this method terminates
+    /// with precisely the reachable arms being reachable - but that problem
+    /// is trivially NP-complete:
+    ///
+    /// ```rust
+    ///     match (var0, var1, var2, var3, ..) {
+    ///         (true, _, _, false, true, ...) => false,
+    ///         (_, true, true, false, _, ...) => false,
+    ///         (false, _, false, false, _, ...) => false,
+    ///         ...
+    ///         _ => true
+    ///     }
+    /// ```
+    ///
+    /// Here the last arm is reachable only if there is an assignment to
+    /// the variables that does not match any of the literals. Therefore,
+    /// compilation would take an exponential amount of time in some cases.
+    ///
+    /// That kind of exponential worst-case might not occur in practice, but
+    /// our simplistic treatment of constants and guards would make it occur
+    /// in very common situations - for example #29740:
+    ///
+    /// ```rust
+    /// match x {
+    ///     "foo" if foo_guard => ...,
+    ///     "bar" if bar_guard => ...,
+    ///     "baz" if baz_guard => ...,
+    ///     ...
+    /// }
+    /// ```
+    ///
+    /// Here we first test the match-pair `x @ "foo"`, which is an `Eq` test.
+    /// It might seem that we would end up with 2 disjoint candidate sets,
+    /// consisting of the first candidate or the other 3, but our algorithm
+    /// doesn't reason about "foo" being distinct from the other constants,
+    /// it considers to latter arms to potentially match after both outcomes,
+    /// which obviously leads to an exponential amount of tests.
```

Comment:
This sentence is long and not fully grammatical. Maybe:

```
+    /// It might seem that we would end up with 2 disjoint candidate sets,
+    /// consisting of the first candidate or the other 3, but our algorithm
+    /// doesn't reason about "foo" being distinct from the other constants;
+    /// instead, it considers the latter arms to potentially match after both outcomes,
+    /// which can lead to an exponential amount of tests.
```

(changes at the end)